broadinstitute/scp-ingest-pipeline
# Parallelizing expression matrix processing for Image Pipeline (SCP-4647)

#### BACKGROUND
Building on the work from SCP-4584, it will be necessary to optimize performance of jobs that process expression data so that they can complete as quickly as possible to allow Image Pipeline to create images faster.  This will reduce the lag between a user uploading an expression matrix and each gene-level entry having a preview image available for viewing.

#### CHANGES
This update introduces several changes.  The largest change is that files are now "chunked" into discrete parts, where the size of each chunk is the total file size divided by the number of available cores on the machine (minus 1).  Each chunk has logic to ensure that it terminates on a newline so as to avoid incomplete data.  Secondly, sparse matrix files now no longer need to be sorted on gene indices before being processed, as the chunks are now processed out into individual gene-level sparse matrix fragments, which each worker process appending to the fragments.  Since these are all single-gene level entries, it does not matter what order the lines are in as the indexes for barcodes remain unchanged.  In addition, missing genes from the sparse matrix (i.e. no observed expression for a given gene from the genes file) will be rendered out as arrays of `0` for each cell in the cluster file.  Lastly, the script `render_expression_arrays.py` has been refactored into the class `ExpressionWriter`, which will allow for easier integration with the rest of the codebase.

Running on a 64 cores GCP machine, this updated script was able to process the 1.3M cell 10X benchmark dataset in 74 minutes (down from ~14 hours for the single-process version).

#### MANUAL TESTING
1. Pull branch and navigate to `scripts/scratch_ingest`
2. Run the following command to process a dense matrix:
```
python3 expression_writer.py --matrix-file ../../tests/data/dense_expression_matrix.txt \ 
                             --cluster-file ../../tests/data/cluster_example.txt \
                             --cluster-name 'Dense Example'
```
3. Confirm that you see messages detailing the chunking logic (note the last chunk has 0 bytes, this is due to the small size of the file and has no impact on output):
```
determining seek points for ../../tests/data/dense_expression_matrix.txt with chunk size 37
reading ../../tests/data/dense_expression_matrix.txt at index 161:264
reading ../../tests/data/dense_expression_matrix.txt at index 265:265
```
4. Confirm that two gene-level JSON arrays are generated - `Dense_Example--Itm2a.json.gz` and `Dense_Example--Sergef.json.gz`
5. Open both files, and confirm that `Itm2a` has all zeros, and `Sergef` has 8 non-zero entries
6. Run the following command to process a sparse matrix:
```
python3 expression_writer.py --matrix-file ../../tests/data/mtx/matrix_with_header.mtx \
                             --genes-file ../../tests/data/mtx/sampled_genes.tsv \
                             --barcodes-file ../../tests/data/mtx/barcodes.tsv \
                             --cluster-file ../../tests/data/mtx/cluster_mtx_barcodes.tsv \
                             --cluster-name 'Sparse Example'
```
7. Confirm chunking statements:
```
determining seek points for ../../tests/data/mtx/matrix_with_header.mtx with chunk size 31
reading ../../tests/data/mtx/matrix_with_header.mtx at index 56:94
reading ../../tests/data/mtx/matrix_with_header.mtx at index 95:135
reading ../../tests/data/mtx/matrix_with_header.mtx at index 136:177
reading ../../tests/data/mtx/matrix_with_header.mtx at index 178:219
reading ../../tests/data/mtx/matrix_with_header.mtx at index 220:220
```
8. In the output directory, confirm there are 7 files, as well as 6 gene-level sparse matrix files in the `gene_entries` folder (there will not be an entry for `HOMER2`)
9. Open `Sparse_Example--HOMER2.json.gz` and confirm it has only zeros, and any of the other files have 2 non-zero entries.
10. (Optional) Re-run any of the above commands with `--precision 1` and confirm that output files have only 1 digit of precision for non-zero values
